### PR TITLE
TP/quantization/weight loading refactor part 2 - Refactor quantized linear logic and extend quantization support to all models

### PR DESCRIPTION
**Pull request by zhuohan123**
Refactor the tensor parallelism, quantization, and weight-loading codes.

Summary of the new features enabled by this PR:
- **All models** are able to be quantized with AWQ and SqueezeLLM, and [soon GPTQ](https://github.com/vllm-project/vllm/pull/1580).
- Model loading code became much simpler.
- Support model parallelism for all MQA/GQA models when the number of key/value heads is smaller than the tensor parallel size.

After this refactor:
- The weight-only quantization configs and their corresponding linear implementation are moved to `vllm/model_executor/layers/quantized_linear/`
- The implementation of quantized linear layers is inherited from `LinearMethodBase` class. Each linear method needs to provide a method to create the weights and a method to apply the weights to an input tensor. Each weight will also include  many attributes 
- The model-parallel linear layers are moved to `vllm/model_executor/layers/linear.py`. They will take the linear method as an argument during initialization. These layers create the weights using the linear method and also implement the weight-loading functions for the weights. The current model-parallel linear layers include:
  - `ReplicatedLinear`: Fully replicated linear.
  - `ColumnParallelLinear`.
  - `RowParallelLinear`.
  - **[New]** `MergedColumnParallelLinear`: Column-parallel linear layer that is a concatenation of multiple Column-parallel linear layers. Follow the same forward logic as `ColumnParallelLinear`, but during weight loading, each sub-weight matrix is sharded and loaded separately.
  - **[New]** `QKVParallelLinear`: Special column-parallel linear layer for QKV transformation. This class handles the weight-loading of group-query attention where the number of key-value heads are different from the query heads.
- Embedding layers and the output logit layers are moved to `vllm/model_executor/layers/vocab_parallel_embedding.py`. A new class `ParallelLMHead` is created for the output logits layer.
- For each model, the weight loading function is drastically simplified. Please refer to the new loading code of each model for details.

The code have been tested for
- LLaMA x (full precision, AWQ, SqueezeLLM) x (with TP, without TP)
- Mistral x (full precision, AWQ) x (with TP, without TP)
- All other models x (with TP, without TP)

TODO:
- [x] Clean up the code
  - [x] Only feed `linear_method` to the model
  - [x] Remove `input_is_parallel=True` and `gather_output=False`
  - [ ] Maybe in another PR: Move `KVCache` to attention.py
- [x] Add code comments
- [x] Merge with the main branch
- [x] Fix the error in #1640

### CODE REVIEW

**Comment by zhuohan123** [comment]
@WoosukKwon This PR is ready for review!

**Comment by void-main** [comment]
Hi @zhuohan123 , what's the expected date for this PR gets merge?

File: vllm/model_executor/layers/quantized_linear/base_config.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/quantized_linear/base_config.py:29]: Actually, we might want to define these methods as instance methods (which takes `self` as the first argument), because their outputs may depend on the configuration. For example, the supported GPUs can be different depending on the quantization bit widths.

File: vllm/model_executor/model_loader.py
Comment (WoosukKwon) [inline vllm/model_executor/model_loader.py:73]: It seems we don't need this now?

File: vllm/model_executor/weight_utils.py
Comment (WoosukKwon) [inline vllm/model_executor/weight_utils.py:266]:
```suggestion
                          loaded_weight:Any, # `torch.Tensor` or `PySafeSlice`,
                          ) -> None:
```

File: vllm/model_executor/layers/quantized_linear/__init__.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/quantized_linear/__init__.py:1]: I'm a bit worried that we are only considering weight-only quantization here. What about changing `quantized_linear` to `quantization` or something like that?

File: vllm/model_executor/utils.py
Comment (WoosukKwon) [inline vllm/model_executor/utils.py:19]: Same for all other places.
```suggestion
    weight_attrs: Optional[Dict[str, Any]],
```

File: vllm/model_executor/utils.py
Comment (WoosukKwon) [inline vllm/model_executor/utils.py:3]:
```suggestion
from typing import Any, Dict, Optional
```

File: vllm/model_executor/layers/linear.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/linear.py:21]:
```suggestion
    """Base class for different (maybe quantized) linear methods."""
```

File: vllm/model_executor/layers/linear.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/linear.py:38]: `FullPrecision` is confusing here because it may mean FP32 specifically.

File: vllm/model_executor/layers/linear.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/linear.py:57]: Don't we need to include the bias term here? While the weight-only quantization methods usually don't quantize the bias, other quantization methods may quantize the bias as well.

File: vllm/model_executor/layers/linear.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/linear.py:214]: `Packed` can be confusing because it may mean the weight tensors are packed.
```suggestion
class MergedColumnParallelLinear(ColumnParallelLinear):
```

**Review by WoosukKwon — COMMENTED** [review]
@zhuohan123 Thanks for the amazing work! I'm generally good with this change, except that some assumptions here will be violated by #1580.

File: vllm/model_executor/layers/linear.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/linear.py:60]: Can we include non-parameter tensors ("buffers" in PyTorch) here?

File: vllm/model_executor/model_loader.py
Comment (zhuohan123) [reply vllm/model_executor/model_loader.py:73]: Fixed!

File: vllm/model_executor/weight_utils.py
Comment (zhuohan123) [reply vllm/model_executor/weight_utils.py:266]: BTW I moved `convert_pyslice_to_tensor` to `hf_model_weights_iterator` in this PR:

https://github.com/vllm-project/vllm/blob/fd4f4d5474ed06e35ae4a0b9de578500c6c5d74d/vllm/model_executor/weight_utils.py#L240

I think the benefit of keeping `PySafeSlice` around is minimal and greatly complicates the code. In this case, we no longer need to deal with `PySafeSlice`. We only load `torch.tensor`.

File: vllm/model_executor/layers/linear.py
Comment (zhuohan123) [reply vllm/model_executor/layers/linear.py:57]: Let's leave this in a future PR. The reason that I don't include bias here is that `RowParallelLinear` applies `all_reduce` before adding bias. Therefore, we cannot directly add bias in the `apply_weights` function.

File: vllm/model_executor/layers/linear.py
Comment (zhuohan123) [reply vllm/model_executor/layers/linear.py:60]: In this case, we need to separate `weights` and `buffers` in `create_weights` function. Let's leave this in a future PR.

**Comment by zhuohan123** [comment]
@WoosukKwon All comments fixed and the PR is ready to be merged!

**Comment by zhuohan123** [comment]
> Hi @zhuohan123 , what's the expected date for this PR gets merge?

Should be merged in a day or two!

File: vllm/model_executor/weight_utils.py
Comment (WoosukKwon) [reply vllm/model_executor/weight_utils.py:266]: Nice. Thanks for the explanation.

**Review by WoosukKwon — APPROVED** [review]
LGTM! Please run some tests (if not exhaustive) before the merge! Thanks a million for the refactoring. Great work!

**Pull request merged**
